docs/build/wasm/from-zero-to-ink-hero/manic-minter/manic-contract.md

```diff
@@ -0,0 +1,141 @@
+---
+sidebar_position: 3
+---
+
+# ManicMinter Contract
+This tutorial will not include complete code from the contract to keep it short and focused on the cross contract calls and e2e tests. The full code for this example is available [here](https://github.com/swanky-dapps/manic-minter)
+
+## Storage
+To start with delete the boilerplate code from the `lib.rs` file.
+The storage will be defined as follows:
+```rust
+pub struct ManicMinter {
+    /// Contract owner
+    owner: AccountId,
+    /// Oxygen contract address
+    token_contract: AccountId,
+    /// Minting price. Caller must pay this price to mint one new token from Oxygen contract
+    price: Balance,
+}
+```
+
+## Error and Types
+We will define the error and types as follows:
+```rust
+/// The ManicMinter error types.
+pub enum Error {
+    /// Returned if not enough balance to fulfill a request is available.
+    BadMintValue,
+    /// The call is not allowed if the caller is not the owner of the contract
+    NotOwner,
+    /// Returned if the token contract account is not set during the contract creation.
+    ContractNotSet,
+    /// Returned if multiplication of price and amount overflows
+    OverFlow,
+}
+
+pub type Result<T> = core::result::Result<T, Error>;
+```
+## Contract Trait
+The following trait will be used to define the contract interface:
+```rust
+pub trait Minting {
+    /// Mint new tokens from Oxygen contract
+    #[ink(message, payable)]
+    fn manic_mint(&mut self, amount: Balance) -> Result<()>;
+
+    /// Set minting price for one Oxygen token
+    #[ink(message)]
+    fn set_price(&mut self, price: Balance) -> Result<()>;
+
+    /// Get minting price for one Oxygen token
+    #[ink(message)]
+    fn get_price(&self) -> Balance;
+}
+```
+
+## Constructor
+The constructor will be defined as follows:
+```rust
+impl ManicMinter {
+    #[ink(constructor)]
+    pub fn new(contract_acc: AccountId) -> Self {
+        Self {
+            owner: Self::env().caller(),
+            token_contract: contract_acc,
+            price: 0,
+        }
+    }
+}
+```
+## Cross Contract Call
+The `manic_mint` method will execute cross contract call to Oxygen contract using `Call Builder`. The method will be defined as follows:
+```rust
+impl Minting for ManicMinter {
+    #[ink(message, payable)]
+    fn manic_mint(&mut self, amount: Balance) -> Result<()> {
+        //---- snip ----
+
+        let _mint_result = build_call::<DefaultEnvironment>()
+            .call(self.token_contract)
+            .gas_limit(5000000000)
+            .exec_input(
+                ExecutionInput::new(Selector::new(ink::selector_bytes!("PSP22Mintable::mint")))
+                    .push_arg(caller)
+                    .push_arg(amount),
+            )
+            .returns::<()>()
+            .try_invoke();
+        Ok(())
```

Comment:
_mint_result is a Result. HEre you return Ok(()) even if the call has failed. 
Maybe you can return an error like [here](https://github.com/paritytech/ink/blob/8bb4b40d76adb28cf046f385cb5842bcd90b8787/integration-tests/multisig/lib.rs#L561)

Reply:
Good point, thanks.
I've also noticed that if error happens in this call, e2e test will crash here and my Error::TransactionFailed will never be returned. I'll open an issue